src/java.base/share/classes/java/util/concurrent/LinkedTransferQueue.java

```diff
@@ -604,45 +619,54 @@ final Object xfer(Object e, long nanos) {
      */
     private void unsplice(DualNode pred, DualNode s) {
-        DualNode sn;
-        if (pred != null && pred.next == s && s != null && (sn = s.next) != s) {
-            if (sn != null)
-                pred.cmpExNext(s, sn);
-            if ((sn == null || !pred.isLive()) &&
-                ((addSweepVote() + 1) & (SWEEP_THRESHOLD - 1)) == 0) {
-                for (DualNode p = head, f, n, u;  // occasionally sweep
-                     p != null && (f = p.next) != null && (n = f.next) != null;)
-                    p = (f.isLive()                   ? f :     // skip
-                         f == p                       ? head :  // stale
-                         f == (u = p.cmpExNext(f, n)) ? n : u); // unspliced
+        boolean seen = false; // try removing by collapsing head
+        for (DualNode h = head, p = h, f; p != null;) {
+            boolean matched;
+            if (p == s)
+                matched = seen = true;
+            else
+                matched = p.matched();
+            if ((f = p.next) == p)
+                p = h = head;
+            else if (f != null && matched)
+                p = f;
+            else {
+                if (p != h && cmpExHead(h, p) == h)
+                    h.next = h; // h.selfLink();
+                break;
+            }
+        }
+        DualNode sn;      // try to unsplice if not pinned
+        if (!seen &&
+            pred != null && pred.next == s && s != null && (sn = s.next) != s &&
+            (sn == null || pred.cmpExNext(s, sn) != s || pred.matched()) &&
+            sweepNow()) { // occasionally sweep if might not have been removed
+            for (DualNode p = head, f, n, u;
+                 p != null && (f = p.next) != null && (n = f.next) != null;) {
+                p = (f == p                       ? head :  // stale
+                     !f.matched()                 ? f :     // skip
+                     f == (u = p.cmpExNext(f, n)) ? n : u); // unspliced
             }
         }
-    }
-
-    /**
-     * Tries to update to new head, forgetting links from previous
-     * head (if it exists) on success.
-     */
-    final DualNode tryAdvanceHead(DualNode h, DualNode p) {
-        DualNode u;
-        if ((u = cmpExHead(h, p)) == h && h != null)
-            h.selfLink();
-        return u;
     }
 
     /**
      * Tries to CAS pred.next (or head, if pred is null) from c to p.
      * Caller must ensure that we're not unlinking the trailing node.
      */
     final boolean tryCasSuccessor(DualNode pred, DualNode c, DualNode p) {
-        // assert p != null && !c.isLive() && c != p;
-        return ((pred != null ?
-                 pred.cmpExNext(c, p) :
-                 tryAdvanceHead(c, p))) == c;
+        // assert p != null && c.matched() && c != p;
+        if (pred != null)
+            return pred.cmpExNext(c, p) == c;
+        else if (cmpExHead(c, p) != c)
+            return false;
+        if (c != null)
```

Comment:
@DougLea If this `if` is not related to the above if-tree, I'd put a newline between them. (Also makes sense to add a newline before the last `return` as well (to indicate that it is standalone from the if)

Reply:
Thanks. Done.